Android 11, Can't access /storage/emulated/0/android/data

App isn't fully capable of the scoped storage API calls for accessing said directory as a file manager.

 - Device: Pixel 3a
 - OS: Android 11
 - Rooted: No
 - Version: 3.5.0

I can't access files saved with apps like Telegram which store downloads in such a location. They don't even show up as recent files in anything else but the built-in file manager in Android (not google files, I have that disabled)

There should be a n actual API call specifically for file managers to access that location but as I'm not completely savvy on the inner workings of modern Android APIs I don't exactly know what it's called. This should probably be looked into heavily when building for Android 11 since the scoped storage API is mandatory now, and it should be seamless to Android 10 users already using the app.

<img src="https://user-images.githubusercontent.com/11806373/96334730-ec19d780-1040-11eb-9954-1c91ddb89881.png" height=500px/> <img src="https://user-images.githubusercontent.com/11806373/96334736-f340e580-1040-11eb-935d-07335a96d4cd.png" height=500px/>

Related crashes

Rename:
```
2020-11-16 03:34:28.529 5634-5935/com.amaze.filemanager.debug E/ActivityThread: Rename recovery failed 
    android.system.ErrnoException: rename failed: EXDEV (Cross-device link)
        at libcore.io.Linux.rename(Native Method)
        at libcore.io.ForwardingOs.rename(ForwardingOs.java:190)
        at libcore.io.BlockGuardOs.rename(BlockGuardOs.java:350)
        at libcore.io.ForwardingOs.rename(ForwardingOs.java:190)
        at android.app.ActivityThread$AndroidOs.rename(ActivityThread.java:7581)
        at java.io.UnixFileSystem.rename(UnixFileSystem.java:368)
        at java.io.File.renameTo(File.java:1412)
        at com.amaze.filemanager.asynchronous.asynctasks.MoveFiles.doInBackground(MoveFiles.java:109)
        at com.amaze.filemanager.asynchronous.asynctasks.MoveFiles.doInBackground(MoveFiles.java:59)
        at android.os.AsyncTask$3.call(AsyncTask.java:394)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:923)

```

Copy:
```
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err: java.io.FileNotFoundException: /storage/emulated/0/Android/data/DummyFile: open failed: EACCES (Permission denied)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:492)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at java.io.FileOutputStream.<init>(FileOutputStream.java:236)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.FileUtil.isWritable(FileUtil.java:754)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.utils.MainActivityHelper.checkFolder(MainActivityHelper.java:420)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.finishCopying(PrepareCopyTask.java:325)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onEndDialog(PrepareCopyTask.java:275)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onEndDialog(PrepareCopyTask.java:269)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onPostExecute(PrepareCopyTask.java:186)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onPostExecute(PrepareCopyTask.java:64)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask.finish(AsyncTask.java:771)
2020-11-16 03:34:56.761 5634-5634/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask.access$900(AsyncTask.java:199)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:788)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:106)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at android.os.Looper.loop(Looper.java:223)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:7656)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err: Caused by: android.system.ErrnoException: open failed: EACCES (Permission denied)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at libcore.io.Linux.open(Native Method)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at libcore.io.ForwardingOs.open(ForwardingOs.java:166)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at libcore.io.BlockGuardOs.open(BlockGuardOs.java:254)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at libcore.io.ForwardingOs.open(ForwardingOs.java:166)
2020-11-16 03:34:56.762 5634-5634/com.amaze.filemanager.debug W/System.err:     at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:7542)
2020-11-16 03:34:56.763 5634-5634/com.amaze.filemanager.debug W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:478)
2020-11-16 03:34:56.763 5634-5634/com.amaze.filemanager.debug W/System.err: 	... 17 more
```

--- a/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
@@ -103,9 +103,11 @@ public class HybridFile {
   public HybridFile(OpenMode mode, String path, String name, boolean isDirectory) {
     this(mode, path);
     this.name = name;
-    if (path.startsWith(SMB_URI_PREFIX) || isSmb()) {
+    if (path.startsWith(SMB_URI_PREFIX) || isSmb() || isDocumentFile() || isOtgFile()) {
       Uri.Builder pathBuilder = Uri.parse(this.path).buildUpon().appendEncodedPath(name);
-      if (isDirectory) pathBuilder.appendEncodedPath("/");
+      if ((path.startsWith(SMB_URI_PREFIX) || isSmb()) && isDirectory) {
+        pathBuilder.appendEncodedPath("/");
+      }
       this.path = pathBuilder.build().toString();
     } else if (path.startsWith(SSH_URI_PREFIX) || isSftp()) {
       this.path += "/" + name;

--- a/app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
@@ -135,9 +135,7 @@ public class Operations {
       @Override
       protected Void doInBackground(Void... params) {
         // checking whether filename is valid or a recursive call possible
-        if (!Operations.isFileNameValid(file.getName(context))
-            && !(file.getMode().equals(OpenMode.OTG)
-                || file.getMode().equals(OpenMode.DOCUMENT_FILE))) {
+        if (!Operations.isFileNameValid(file.getName(context))) {
           errorCallBack.invalidName(file);
           return null;
         }
@@ -280,9 +278,7 @@ public class Operations {
       @Override
       protected Void doInBackground(Void... params) {
         // check whether filename is valid or not
-        if (!Operations.isFileNameValid(file.getName(context))
-            && !(file.getMode().equals(OpenMode.OTG)
-                || file.getMode().equals(OpenMode.DOCUMENT_FILE))) {
+        if (!Operations.isFileNameValid(file.getName(context))) {
           errorCallBack.invalidName(file);
           return null;
         }
@@ -449,10 +445,7 @@ public class Operations {
       protected Void doInBackground(Void... params) {
         // check whether file names for new file are valid or recursion occurs.
         // If rename is on OTG, we are skipping
-        if (!(oldFile.getMode().equals(newFile.getMode())
-                && (oldFile.getMode().equals(OpenMode.OTG)
-                    || oldFile.getMode().equals(OpenMode.DOCUMENT_FILE)))
-            && !Operations.isFileNameValid(newFile.getName(context))) {
+        if (!Operations.isFileNameValid(newFile.getName(context))) {
           errorCallBack.invalidName(newFile);
           return null;
         }

--- a/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/fragments/MainFragment.java
@@ -1312,11 +1312,7 @@ public class MainFragment extends Fragment
                   .rename(
                       mainFragmentViewModel.getOpenMode(),
                       f.getPath(),
-                      Uri.parse(mainFragmentViewModel.getCurrentPath())
-                          .buildUpon()
-                          .appendEncodedPath(name1)
-                          .build()
-                          .toString(),
+                      mainFragmentViewModel.getCurrentPath(),
                       name1,
                       f.isDirectory(),
                       getActivity(),
@@ -1414,7 +1410,7 @@ public class MainFragment extends Fragment
               loadlist(currentFile.getPath(), false, mainFragmentViewModel.getOpenMode());
             } else {
               List<String> pathSegments = Uri.parse(currentFile.getPath()).getPathSegments();
-              if (pathSegments.size() < 4) {
+              if (pathSegments.size() < 3) {
                 mainFragmentViewModel.setOpenMode(OpenMode.FILE);
                 String subPath = pathSegments.get(1);
                 currentFile.setMode(OpenMode.FILE);

--- a/app/src/main/java/com/amaze/filemanager/utils/MainActivityHelper.java
+++ b/app/src/main/java/com/amaze/filemanager/utils/MainActivityHelper.java
@@ -68,7 +68,6 @@ import android.content.BroadcastReceiver;
 import android.content.Context;
 import android.content.Intent;
 import android.content.SharedPreferences;
-import android.net.Uri;
 import android.os.AsyncTask;
 import android.os.Build;
 import android.os.Bundle;
@@ -161,15 +160,7 @@ public class MainActivityHelper {
           EditText textfield = dialog.getCustomView().findViewById(R.id.singleedittext_input);
           mkDir(
               new HybridFile(openMode, path),
-              new HybridFile(
-                  openMode,
-                  Uri.parse(path)
-                      .buildUpon()
-                      .appendEncodedPath(textfield.getText().toString().trim())
-                      .build()
-                      .toString(),
-                  textfield.getText().toString().trim(),
-                  true),
+              new HybridFile(openMode, path, textfield.getText().toString().trim(), true),
               ma);
           dialog.dismiss();
         },
@@ -202,15 +193,7 @@ public class MainActivityHelper {
           EditText textfield = dialog.getCustomView().findViewById(R.id.singleedittext_input);
           mkFile(
               new HybridFile(openMode, path),
-              new HybridFile(
-                  openMode,
-                  Uri.parse(path)
-                      .buildUpon()
-                      .appendEncodedPath(textfield.getText().toString().trim())
-                      .build()
-                      .toString(),
-                  textfield.getText().toString().trim(),
-                  false),
+              new HybridFile(openMode, path, textfield.getText().toString().trim(), false),
               ma);
           dialog.dismiss();
         },

